Unify semantic errors with standard errors

martint opened the pull request:
Semantic errors were introduced before PrestoException and friends existed.
When PrestoException and error codes were added, we ended up with two classes
of errors. All semantic errors were being bundled under a single top-level
error, which made it hard to propagate details about such errors to client,
event listeners, etc.

This change gets rid of SemanticException and related classes, and flattens
the error codes into StandardErrorCode.

martint committed:
Unify semantic errors with standard errors

Semantic errors were introduced before PrestoException and friends existed.
When PrestoException and error codes were added, we ended up with two classes
of errors. All semantic errors were being bundled under a single top-level
error, which made it hard to propagate details about such errors to client,
event listeners, etc.

This change gets rid of SemanticException and related classes, and flattens
the error codes into StandardErrorCode.

electrum commented on presto-main/src/test/java/io/prestosql/sql/analyzer/TestAnalyzer.java:
```diff
@@ -1711,45 +1700,45 @@ private void analyze(Session clientSession, @Language("SQL") String query)
-                NodeLocation expected = location.get();
-                NodeLocation actual = e.getNode().getLocation().get();
+                Location expected = location.get();
+                Location actual = e.getLocation().get();
```
This can throw. Change to something like
```java
Location actual = e.getLocation().orElseThrow(() ->
        new AssertionError("Expected location but actual is missing"));

electrum commented on presto-main/src/test/java/io/prestosql/sql/analyzer/TestAnalyzer.java:
```diff
@@ -1759,15 +1748,15 @@ private void assertFails(Session session, SemanticErrorCode error, Optional<Node
-            if (e.getCode() != error) {
-                fail(format("Expected error %s, but found %s: %s", error, e.getCode(), e.getMessage()), e);
+        catch (PrestoException e) {
+            if (e.getErrorCode() != error.toErrorCode()) {
```
This will need `equals` with `ErrorCodeSupplier`

electrum commented on presto-spi/src/main/java/io/prestosql/spi/StandardErrorCode.java:
```diff
@@ -64,6 +64,47 @@
     INVALID_SPATIAL_PARTITIONING(0x0000_0029, USER_ERROR),
     INVALID_ANALYZE_PROPERTY(0x0000_002A, USER_ERROR),
     TYPE_NOT_FOUND(0x0000_002B, USER_ERROR),
+    CATALOG_NOT_FOUND(0x0000_002C, USER_ERROR),
```
As a pre-commit, we might want to change these to decimals, similar to what we now do for connectors:
```java
GENERIC_USER_ERROR(userCode(0), USER_ERROR),
SYNTAX_ERROR(userCode(1), USER_ERROR),
ABANDONED_QUERY(userCode(2), USER_ERROR),

private static int userCode(int offset)
{
    return offset; // 0x0000_0000
}

private static int internalCode(int offset)
{
    return 0x0001_0000 + offset;
}

private static int resourcesCode(int offset)
{
    return 0x0002_0000 + offset;
}

martint replied:
I don't think we should rely on specific "offsets" to indicate the different types of errors. I understand that's how we did it originally, but the error category is more robust. Error codes should be opaque and we shouldn't ascribe any specific meaning to which range an error falls in. It's also easier to search the codebase for the corresponding error when it appears as a literal instead of being derived programmatically. I do agree that we should display them as simple integers instead of hex, though.

electrum replied:
Maybe “offset” is the wrong name for the parameter. My point is we can replace the hex values with numbers. But we need to preserve the existing mapping and this is an easy way to do it. But a straight conversion from hex to decimal is bad because the higher value ones will have a seemingly random value (which is only understandable in hex).

electrum commented on presto-spi/src/main/java/io/prestosql/spi/StandardErrorCode.java:
```diff
@@ -64,6 +64,47 @@
+    TABLE_NOT_FOUND(0x0000_002E, USER_ERROR),
+    COLUMN_NOT_FOUND(0x0000_002F, USER_ERROR),
+    ROLE_NOT_FOUND(0x0000_0030, USER_ERROR),
+    SCHEMA_ALREADY_EXISTS(0x0000_0031, USER_ERROR),
```
We need to cleanup the existing usages of `ALREADY_EXISTS` in the future

electrum commented on presto-spi/src/main/java/io/prestosql/spi/StandardErrorCode.java:
```diff
@@ -64,6 +64,47 @@
+    TABLE_ALREADY_EXISTS(0x0000_0032, USER_ERROR),
+    COLUMN_ALREADY_EXISTS(0x0000_0033, USER_ERROR),
+    ROLE_ALREADY_EXISTS(0x0000_0034, USER_ERROR),
+    NAMED_QUERY_ALREADY_EXISTS(0x0000_0035, USER_ERROR),
```
This name is confusing since it's similar to the others, but means something different. How about `DUPLICATE_NAMED_QUERY`, similar to `DUPLICATE_PROPERTY`

electrum commented on presto-spi/src/main/java/io/prestosql/spi/StandardErrorCode.java:
```diff
@@ -64,6 +64,47 @@
+    MISSING_COLUMN_NAME(0x0000_0036, USER_ERROR),
+    MISSING_CATALOG_NAME(0x0000_0037, USER_ERROR),
+    MISSING_SCHEMA_NAME(0x0000_0038, USER_ERROR),
+    INVALID_SCHEMA_NAME(0x0000_0039, USER_ERROR),
```
This is only used in one place, `MetadataUtil`, as a helper method for the `QualifiedName` passed to schema statements like `CREATE SCHEMA`. This can probably use `SYNTAX_ERROR`, like the related `Too many dots in table name` error.

electrum commented on presto-spi/src/main/java/io/prestosql/spi/StandardErrorCode.java:
```diff
@@ -64,6 +64,47 @@
+    EXPRESSION_NOT_CONSTANT(0x0000_004B, USER_ERROR),
+    INVALID_ARGUMENTS(0x0000_004C, USER_ERROR),
+    TOO_MANY_ARGUMENTS(0x0000_004D, USER_ERROR),
+    INVALID_TRANSACTION_MODE(0x0000_004E, USER_ERROR),
```
This can probably be `SYNTAX_ERROR`. It's only used when multiple isolation levels or read modes are specified, which is effectively a syntax error.

electrum commented on presto-spi/src/main/java/io/prestosql/spi/StandardErrorCode.java:
```diff
@@ -64,6 +64,47 @@
+    INVALID_PARAMETER_USAGE(0x0000_0051, USER_ERROR),
+    VIEW_IS_STALE(0x0000_0052, USER_ERROR),
+    VIEW_IS_RECURSIVE(0x0000_0053, USER_ERROR),
+    VIEW_ANALYSIS_ERROR(0x0000_0054, USER_ERROR),
```
Not needed after switching to `INVALID_VIEW`, as explained in other comments.

martint commented on presto-main/src/main/java/io/prestosql/execution/AddColumnTask.java:
```diff
@@ -80,16 +80,16 @@ public String getName()
         }
         if (type.equals(UNKNOWN)) {
-            throw new SemanticException(TYPE_MISMATCH, element, "Unknown type '%s' for column '%s'", element.getType(), element.getName());
+            throw semanticException(TYPE_MISMATCH, element, "Unknown type '%s' for column '%s'", element.getType(), element.getName());
```
`COLUMN_TYPE_UNKNOWN` is actually more appropriate here.

martint commented on presto-spi/src/main/java/io/prestosql/spi/PrestoException.java:
```diff
@@ -13,34 +13,59 @@
+    {
+        String message = getRawMessage();
+        if (location.isPresent()) {
+            message = format("line %s:%s: %s", location.get().getLineNumber(), location.get().getColumnNumber(), message);
```
I don't want to rely on a toString that might be used for debugging (i.e., the format may be appropriate for showing in the debugger but not for presenting to end users.

martint commented on presto-spi/src/main/java/io/prestosql/spi/PrestoException.java:
```diff
@@ -13,34 +13,59 @@
 
     public PrestoException(ErrorCodeSupplier errorCode, String message)
     {
-        this(errorCode, message, null);
```
Ah, yes, leftover from intermediate changes.

martint commented on presto-spi/src/main/java/io/prestosql/spi/Location.java:
```diff
@@ -0,0 +1,50 @@
+    private final int lineNumber;
+    private final int columnNumber;
+
+    @JsonCreator
```
Oops, forgot to remove this when I cloned the ErrorLocation class from the client.

martint commented on presto-main/src/main/java/io/prestosql/sql/analyzer/StatementAnalyzer.java:
```diff
@@ -2002,7 +1997,7 @@ private RelationType analyzeView(Query query, QualifiedObjectName name, Optional
             catch (RuntimeException e) {
                 throwIfInstanceOf(e, PrestoException.class);
-                throw new SemanticException(VIEW_ANALYSIS_ERROR, node, "Failed analyzing stored view '%s': %s", name, e.getMessage());
+                throw semanticException(VIEW_ANALYSIS_ERROR, node, "Failed analyzing stored view '%s': %s", name, e.getMessage());
```
Hmm.. I realized I'll need to adjust this `catch` block, too, since it rethrows PrestoException. Now every semantic error will be a PrestoException, so it will almost never get to this line.

martint commented on presto-main/src/main/java/io/prestosql/sql/analyzer/StatementAnalyzer.java:
```diff
@@ -2176,10 +2171,10 @@ private void analyzeOffset(Offset node)
             }
             catch (NumberFormatException e) {
-                throw new SemanticException(INVALID_OFFSET_ROW_COUNT, node, "Invalid OFFSET row count: %s", node.getRowCount());
+                throw semanticException(TYPE_MISMATCH, node, "Invalid OFFSET row count: %s", node.getRowCount());
```
I went back and forth on this when I made the change. I came to the conclusion TYPE_MISMATCH was more appropriate for consistency with what happens when a large value (e.g., bigint) is passed to a function that takes a smaller type (e.g., integer). E.g.,

```
presto> select to_big_endian_32(999999999999999999);
Query 20190720_034047_00006_ajbhr failed: line 1:8: Unexpected parameters (bigint) for function to_big_endian_32. Expected: to_big_endian_32(integer)
```

martint commented on presto-main/src/main/java/io/prestosql/sql/analyzer/ExpressionAnalyzer.java:
```diff
@@ -858,24 +858,39 @@ protected Type visitFunctionCall(FunctionCall node, StackableAstVisitorContext<C
+                function = metadata.resolveFunction(node.getName(), argumentTypes);
+            }
+            catch (PrestoException e) {
+                if (e.getLocation().isPresent()) {
```
No. This is a tricky one. `metadata.resolveFunction` can fail either because the function doesn't exist (which throws a custom PrestoException w/o a location, or because the analysis of `argumentTypes` fails. The `argumentTypes` are analyzed lazily to deal with lambda expressions. In that case, we just want to re-throw the original exception.

electrum replied:
Ah, I didn't realize `TypeSignatureProvider` was lazy. So it does indeed have AST nodes inside.

martint commented on presto-main/src/main/java/io/prestosql/metadata/AbstractPropertyManager.java:
```diff
@@ -96,20 +95,7 @@ public final void removeProperties(CatalogName catalogName)
-            catch (SemanticException e) {
-                throw new PrestoException(
-                        propertyError,
-                        format("Invalid value for %s property '%s': Cannot convert [%s] to %s",
```
Let me look at this again. If I recall, there was some issue with some tests relying on the underlying PrestoException and message being thrown, which is no longer possible to discriminate now that everything is a PrestoException.

martint replied:
I restored this catch block. I had to adjust the expected message in one of the tests.

martint committed:
Use decimal error codes

Originally, the error category was encoded in the error code number. We've since
switched to explicitly tagging the error codes with the category they belong to,
so the "shape" of the numeric code no longer means anything. This change makes
them plain decimal numbers to better convey that fact.

martint commented on presto-main/src/main/java/io/prestosql/sql/analyzer/ExpressionAnalyzer.java:
```diff
@@ -652,19 +652,19 @@ protected Type visitSubscriptExpression(SubscriptExpression node, StackableAstVi
                 List<Type> rowTypes = baseType.getTypeParameters();
                 if (indexValue > rowTypes.size()) {
-                    throw new SemanticException(INVALID_PARAMETER_USAGE, node.getIndex(), "Subscript index out of bounds: %s, max value is %s", indexValue, rowTypes.size());
+                    throw semanticException(NUMERIC_VALUE_OUT_OF_RANGE, node.getIndex(), "Subscript index out of bounds: %s, max value is %s", indexValue, rowTypes.size());
```
Hmm... yes and no. From the point of view of the `[]` operator, it's an invalid argument. A more appropriate value would be INVALID_FUNCTION_ARGUMENT, like in the corresponding operator for the ARRAY and MAP types.

electrum commented on presto-main/src/main/java/io/prestosql/metadata/AbstractPropertyManager.java:
```diff
@@ -110,6 +109,7 @@ public final void removeProperties(CatalogName catalogName)
                                 property.getSqlType()),
                         e);
             }
+//            Object sqlObjectValue = evaluatePropertyValue(sqlProperty.getValue(), property.getSqlType(), session, metadata, parameters);
```
I assume you meant to delete this

martint replied:
Yes, of course :)

electrum commented on presto-main/src/main/java/io/prestosql/sql/analyzer/ExpressionAnalyzer.java:
```diff
@@ -858,24 +859,44 @@ protected Type visitFunctionCall(FunctionCall node, StackableAstVisitorContext<C
+            }
+            catch (PrestoException e) {
+                if (e.getLocation().isPresent()) {
+                    // If analysis of any of the argument types (which is done lazily to deal with lambda expressions) fails, we
```
Nit: these comments would look better if the text was balanced over the two lines

electrum commented on presto-main/src/main/java/io/prestosql/sql/analyzer/ExpressionAnalyzer.java:
```diff
@@ -858,24 +859,44 @@ protected Type visitFunctionCall(FunctionCall node, StackableAstVisitorContext<C
+
+                // otherwise, it must have failed due to a missing function or other reason, so we report an error at the
+                // current location
+                throw new PrestoException(
```
Maybe introduce a variable to avoid excessive wrapping (also makes the throw easier to read)
```java
Location location = node.getLocation()
        .map(location -> new Location(location.getLineNumber(), location.getColumnNumber()));
throw new PrestoException(e::getErrorCode, location, e.getMessage(), e);

martint replied:
(It's actually an optional, so it ends up being slightly more verbose):

```java
Optional<Location> errorLocation = node.getLocation()
        .map(location -> new Location(location.getLineNumber(), location.getColumnNumber()));
throw new PrestoException(e::getErrorCode, errorLocation, e.getMessage(), e);
```

I'm not sure that ends up being much better. Yes, it removes wrapping, but there's an additional variable to think about (and the fact it's there makes it seem "important"). A better way would be to introduce a method for that translation:

```java
throw new PrestoException(e::getErrorCode, extractLocation(node), e.getMessage(), e);
```

electrum commented on presto-main/src/main/java/io/prestosql/sql/analyzer/SemanticExceptions.java:
```diff
@@ -13,39 +13,63 @@
+
+    public static PrestoException semanticException(ErrorCodeSupplier code, Node node, String format, Object... args)
+    {
+        throw new PrestoException(
```
How about
```java
Location errorLocation = node.getLocation()
        .map(location -> new Location(location.getLineNumber(), location.getColumnNumber()));
throw new PrestoException(code, errorLocation, format(format, args), null);

electrum commented on presto-spi/src/main/java/io/prestosql/spi/StandardErrorCode.java:
```diff
@@ -20,129 +20,129 @@
     /**/;
 
-    // Connectors can use error codes starting at the range 0x0100_0000
+    // Connectors can use error codes starting at 16777216
```
Keep this as hex, since we do assignments to connectors using 2^16 hex blocks (see wiki page)
